---
title: "DEG_tumor_type"
author: "Christoph Luh"
date: "2025-11-08"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

---
title: "DESeq2_expression"
author: "Christoph Luh"
date: "2025-11-07"
output: html_document
---

The aim of this Rmd is to do a Differential Expression Analysis.
Especially the differences between the astrocytoma types is of great
interest. Till the lines "\_\_\_\_\_\_" data is only loaded. You can
skip that part.

# Data Loading ##All the data is stored on the main branch

```{bash}
git checkout main
```

## Load and filter the meta data

```{r}
#Read meta_data
meta_data <- read.csv("meta_data_unfiltered", row.names = 1)
meta_data <- data.frame(Run = meta_data$Run, history = meta_data$history)

#Remove unwanted histroies (only astrocytomas are left now)
exclude_terms <- c(
  "recurrent astrocytomas",
  "secondary Glioblastomas",
  "recurrent anaplastic oligodendrogliomas",
  "recurrent anaplastic astrocytomas",
  "recurrent anaplastic oligodendroastrocytomas",
  "recurrent Glioblastomas",
  "recurrent oligodendroastrocytomas",
  "anaplastic oligodendroastrocytomas", 
  "anaplastic oligodendrogliomas",
  "oligodendrogliomas",
  "oligodendroastrocytomas"                           
)

meta_data <- meta_data[!(meta_data$history %in% exclude_terms), ]

#An additional column called "severity" is added. The astrocytomas are graded for high and low grade. All tumors starting with anaplastic or primary are graded as high, the other ones as low.
meta_data$severity <- ifelse(
  grepl("anaplastic|primary", meta_data$history, ignore.case = TRUE),
  "high",
  "low"
)

#Create filtering columns --> In each column only the rows with the corresponding tumor are rated as 1, the other ones as 0
tumor_types <- c("astrocytomas", "oligodendroastrocytomas", "oligodendrogliomas")

for (tumor in tumor_types) {
  col_name <- tumor
  
  meta_data[[col_name]] <- ifelse(
    meta_data$history %in% c(tumor, paste("anaplastic", tumor)),
    1,
    0
  )
}

meta_data$glioblastomas <- ifelse(
  meta_data$history %in% c("primary Glioblastomas"),
  1,
  0
)

print(meta_data)

```

## Load and filter expression data

```{r}
expr <- read.csv("expression_matrix.csv", row.names = 1)
expr <- expr[colnames(expr) %in% meta_data$Run]
expr
```

## Load the genes meta data from ENSEMBL

```{r}
library(biomaRt)
library(dplyr)


ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl")
meta_genes <- getBM(attributes = c("ensembl_gene_id",
                                   "ensembl_gene_id_version",
                                   "hgnc_symbol",
                                   "description",
                                   "chromosome_name",
                                   "start_position",
                                   "end_position",
                                   "strand"),
                    filters = "ensembl_gene_id_version",
                    values = rownames(expr),
                    mart = ensembl) %>%
  right_join(data.frame(ensembl_gene_id_version = rownames(expr)),
             by = "ensembl_gene_id_version") %>%
  distinct(ensembl_gene_id_version, .keep_all = TRUE)


expr <- expr[meta_genes$ensembl_gene_id_version,]

```

## Filter for expressed genes

```{r}
library(ggplot2)
num_det <- rowSums(expr > 0)
hist(num_det)
expressed <- rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1
meta_genes$expressed <- expressed
```

## Load DDS File

```{r}
library(DESeq2)
dds <- readRDS("dds.rds")
```

## Clean the colData(dds) from all the secondary/recursive tumors by using meta_data

```{r}
# Filter samples in dds which are not in meta_data
dds <- dds[, colnames(dds) %in% meta_data$Run]

# Get colData(dds) into the right order like in meta_data
dds <- dds[, match(meta_data$Run, colnames(dds))]

# Check if all dds rows are the same as in meta_data, should be TRUE
all(colnames(dds) == meta_data$Run)

```

## Factorize History

```{r}
#The columns in the dds file used as a design, must be a factor
colData(dds)$history <- as.factor(colData(dds)$history)
```

## Remove unused histories (No factor should have 0 usages in the table)

```{r}
colData(dds)$history <- droplevels(colData(dds)$history)
levels(colData(dds)$history)   
table(colData(dds)$history)    

```

# Differential Expression Analysis

## DEG 1: Comparing different histories with another

```{r}
#Set astrocytomas as a reference
colData(dds)$history <- relevel(colData(dds)$history, ref = "astrocytomas")
levels(colData(dds)$history)
#Set history as the design criteria
design(dds) <- ~ history
#Run teh DESeq
dds_history <- DESeq(dds)
```

# Comparison of different histories

This chunk creates results variables for all different combinations of
history comparisons and filters out the significantly differentially
expressed genes (padj \< 0.05)

This chunk also adds a column called "SYMBOL", which has information
about the gene that is encoded by this cryptic ENSEBL ID.

```{r}
#Low Grade Astrocytoma vs. Anaplastic Astrocytoma
res_LGA_APA <- as.data.frame(results(dds_history, contrast = c("history", "astrocytomas", "anaplastic astrocytomas")))
genes_clean <- sub("\\.\\d+$", "", rownames(res_LGA_APA))
res_LGA_APA$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
res_LGA_APA$clean_names <- genes_clean
sig_LGA_APA <- res_LGA_APA[!is.na(res_LGA_APA$padj) & res_LGA_APA$padj < 0.05, ]

#Low Grade Astrocytoma vs. Primary Glioblastoma
res_LGA_GBM <- as.data.frame(results(dds_history, contrast = c("history", "astrocytomas", "primary Glioblastomas")))
genes_clean <- sub("\\.\\d+$", "", rownames(res_LGA_GBM))
res_LGA_GBM$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
res_LGA_GBM$clean_names <- genes_clean
sig_LGA_GBM <- res_LGA_GBM[!is.na(res_LGA_GBM$padj) & res_LGA_GBM$padj < 0.05, ]

#Low Anaplastic Astrocytoma vs. Primary Glioblastoma
res_APA_GBM <- as.data.frame(results(dds_history, contrast = c("history", "anaplastic astrocytomas", "primary Glioblastomas")))
genes_clean <- sub("\\.\\d+$", "", rownames(res_APA_GBM))
res_APA_GBM$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
res_APA_GBM$clean_names <- genes_clean
sig_APA_GBM <- res_APA_GBM[!is.na(res_APA_GBM$padj) & res_APA_GBM$padj < 0.05, ]
```

```{r}
library(org.Hs.eg.db)
library(clusterProfiler)
library(EnhancedVolcano)
library(AnnotationDbi)

```

## Low-grade Astrocytoma vs. Anaplastic Astrocytoma

*Significant DEGs*

```{r}
class(sig_LGA_APA)
sig_LGA_APA[, c("clean_names", "log2FoldChange", "padj", "symbol")]

class(sig_LGA_GBM)
sig_LGA_GBM[, c("clean_names", "log2FoldChange", "padj", "symbol")]

class(sig_APA_GBM)
sig_APA_GBM[, c("clean_names", "log2FoldChange", "padj", "symbol")]
```

*Volcano Plot*

```{r, fig.height=10,fig.width=12}
EnhancedVolcano(res_LGA_APA, x = "log2FoldChange", y = "padj", lab = res_LGA_APA$symbol, pCutoff = 0.05)
EnhancedVolcano(res_LGA_GBM, x = "log2FoldChange", y = "padj", lab = res_LGA_GBM$symbol, pCutoff = 0.05)
EnhancedVolcano(res_APA_GBM, x = "log2FoldChange", y = "padj", lab = res_APA_GBM$symbol, pCutoff = 0.05)
```

*Enrichment Analysis* Creation of Background for enrichtment

```{r}
background <- meta_genes[meta_genes$expressed, "ensembl_gene_id"]
```

Analysis for the different combinations of histories

```{r}
#Do the enrichment analysis
GO_LGA_APA <- enrichGO(gene = sig_LGA_APA$clean_names, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP", universe = background)
GO_LGA_GBM <- enrichGO(gene = sig_LGA_GBM$clean_names, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP", universe = background)
#Gene enrichment for APA vs. GMB is not sensible, as we only have three genes
```

```{r}
#Dataframe of the enrighment GO analysis
GO_LGA_APA.df <- as.data.frame(GO_LGA_APA[GO_LGA_APA$p.adjust < 0.005])
GO_LGA_APA.df

GO_LGA_GBM.df <- as.data.frame(GO_LGA_GBM[GO_LGA_GBM$p.adjust < 0.005])
GO_LGA_GBM.df
```

*Barplot of the gene enrichment analysis*

```{r, fig.height=10,fig.width=12}
barplot(GO_LGA_APA,
      x = "GeneRatio",
      color = "p.adjust",
      title = "GO Enrichment",
      showCategory = 5,
      label_format = 80
)

barplot(GO_LGA_GBM,
      x = "GeneRatio",
      color = "p.adjust",
      title = "GO Enrichment",
      showCategory = 13,
      label_format = 80
)
```
