---
title: "DEG_tumor_type"
author: "Christoph Luh"
date: "2025-11-08"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

---
title: "DESeq2_expression"
author: "Christoph Luh"
date: "2025-11-07"
output: html_document
---

The aim of this Rmd is to do a Differential Expression Analysis.
Especially the differences between the astrocytoma types is of great
interest. Till the lines "\_\_\_\_\_\_" data is only loaded. You can
skip that part.

# Data Loading ##All the data is stored on the main branch

```{bash}
git checkout main
```

## Load and filter the meta data ##

```{r}
#Read meta_data
meta_data <- read.csv("meta_data_unfiltered", row.names = 1)
meta_data <- data.frame(Run = meta_data$Run, history = meta_data$history)

#Remove unwanted histroies (only astrocytomas are left now)
exclude_terms <- c(
  "recurrent astrocytomas",
  "secondary Glioblastomas",
  "recurrent anaplastic oligodendrogliomas",
  "recurrent anaplastic astrocytomas",
  "recurrent anaplastic oligodendroastrocytomas",
  "recurrent Glioblastomas",
  "recurrent oligodendroastrocytomas",
  "anaplastic oligodendroastrocytomas", 
  "anaplastic oligodendrogliomas",
  "oligodendrogliomas",
  "oligodendroastrocytomas"                           
)

meta_data <- meta_data[!(meta_data$history %in% exclude_terms), ]

#An additional column called "severity" is added. The astrocytomas are graded for high and low grade. All tumors starting with anaplastic or primary are graded as high, the other ones as low.
meta_data$severity <- ifelse(
  grepl("anaplastic|primary", meta_data$history, ignore.case = TRUE),
  "high",
  "low"
)

#Create filtering columns --> In each column only the rows with the corresponding tumor are rated as 1, the other ones as 0
tumor_types <- c("astrocytomas", "oligodendroastrocytomas", "oligodendrogliomas")

for (tumor in tumor_types) {
  col_name <- tumor
  
  meta_data[[col_name]] <- ifelse(
    meta_data$history %in% c(tumor, paste("anaplastic", tumor)),
    1,
    0
  )
}

meta_data$glioblastomas <- ifelse(
  meta_data$history %in% c("primary Glioblastomas"),
  1,
  0
)

print(meta_data)

```

## Load and filter expression data ##

```{r}
expr <- read.csv("expression_matrix.csv", row.names = 1)
expr <- expr[colnames(expr) %in% meta_data$Run]
expr
```

## Load the genes meta data from ENSEMBL ##

```{r}
library(biomaRt)
library(dplyr)


ensembl <- useEnsembl(biomart = "ensembl",
                      dataset = "hsapiens_gene_ensembl")
meta_genes <- getBM(attributes = c("ensembl_gene_id",
                                   "ensembl_gene_id_version",
                                   "hgnc_symbol",
                                   "description",
                                   "chromosome_name",
                                   "start_position",
                                   "end_position",
                                   "strand"),
                    filters = "ensembl_gene_id_version",
                    values = rownames(expr),
                    mart = ensembl) %>%
  right_join(data.frame(ensembl_gene_id_version = rownames(expr)),
             by = "ensembl_gene_id_version") %>%
  distinct(ensembl_gene_id_version, .keep_all = TRUE)


expr <- expr[meta_genes$ensembl_gene_id_version,]

```

## Filter for expressed genes ## 

```{r}
library(ggplot2)
num_det <- rowSums(expr > 0)
hist(num_det)
expressed <- rowMeans(expr > 0) >= 0.5 | rowMeans(expr) >= 1
meta_genes$expressed <- expressed
```

## Load DDS File ##

```{r}
library(DESeq2)
dds <- readRDS("dds.rds")
```

## Clean the colData(dds) from all the secondary/recursive tumors by using meta_data ##

```{r}
# Filter samples in dds which are not in meta_data
dds <- dds[, colnames(dds) %in% meta_data$Run]

# Get colData(dds) into the right order like in meta_data
dds <- dds[, match(meta_data$Run, colnames(dds))]

# Check if all dds rows are the same as in meta_data, should be TRUE
all(colnames(dds) == meta_data$Run)

```

## Factorize History ##

```{r}
#The columns in the dds file used as a design, must be a factor
colData(dds)$history <- as.factor(colData(dds)$history)
```

## Remove unused histories (No factor should have 0 usages in the table) ##

```{r}
colData(dds)$history <- droplevels(colData(dds)$history)
levels(colData(dds)$history)   
table(colData(dds)$history)    

```

# Differential Expression Analysis ##DEG 1: Comparing different historys #
with another

```{r}
#Set astrocytomas as a reference
colData(dds)$history <- relevel(colData(dds)$history, ref = "astrocytomas")
levels(colData(dds)$history)
#Set history as the design criteria
design(dds) <- ~ history
#Run teh DESeq
dds_history <- DESeq(dds)
```

# Ignore this part #
# _____________________________ #

## DEG 2: comparing low grade astrocytomas to high grade astrocytomas ##
(anapl. astrocyt. + prim. Glioblast.)

```{r}
#Create a column in dds for the combined histories
colData(dds)$history_combined <- as.character(colData(dds)$history)
colData(dds)$history_combined[
  colData(dds)$history_combined %in% c("anaplastic astrocytomas", "primary Glioblastomas")
] <- "high_grade"
colData(dds)$history_combined[
  colData(dds)$history_combined == "astrocytomas"
] <- "low_grade"
colData(dds)$history_combined <- factor(colData(dds)$history_combined)

# Set design and run
design(dds) <- ~ history_combined
dds_loast_hiast <- DESeq(dds)

```

```{r}
vsdata <- vst(dds_history, blind = FALSE)
plotPCA(vsdata, intgroup = "history")
```

```{r}
#Create a result variable that compares low_grade to high_grade
res_low_vs_high <- results(dds_loast_hiast, contrast = c("history_combined", "low_grade", "high_grade"))
```

```{r}
#This chunk adds the gene names of the genes at the last column. As the ENSEMBL gene IDs do not really explain what gene this is, this columns is useful for understanding
library(org.Hs.eg.db)
res_low_vs_high.df <- as.data.frame(res_low_vs_high)
genes_clean <- sub("\\.\\d+$", "", rownames(res_low_vs_high.df))
res_low_vs_high.df$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
res_low_vs_high.df
```

## Volcanoplot of the DEG 2 (High vs. Low level) ##

```{r, fig.height=10,fig.width=10}
library(EnhancedVolcano)
EnhancedVolcano(res_low_vs_high.df, x = "log2FoldChange", y = "padj", lab = res_low_vs_high.df$symbol)
```

No significant DEGs... Uff...
# _____________________________________ #

## With the results file one can create different comparisons of certain ##
astrocytoma types. Here ATM anapalstic astocytoma vs. primary
Glioblastoma

```{r}
res_astro <- results(dds_history, contrast = c("history", "astrocytomas", "anaplastic astrocytomas"))

#This chunk adds the gene names of the genes at the last column. As the ENSEMBL gene IDs do not really explain what gene this is, this columns is useful for understanding
library(org.Hs.eg.db)
res_astro.df <- as.data.frame(res_astro)
genes_clean <- sub("\\.\\d+$", "", rownames(res_astro.df))
res_astro.df$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
res_astro.df
```

# Visualization of the DEG ##Volcano Plot of the DESeq History #

```{r, fig.height=10,fig.width=12}
library(EnhancedVolcano)
EnhancedVolcano(res_astro.df, x = "log2FoldChange", y = "padj", lab = res_low_vs_high.df$symbol)
```

# Comparison of different histories This chunk creates results variables #
for all different combinations of history comparisons and filters out
the significantly differentially expressed genes (padj \< 0.05)

```{r}
#Low Grade Astrocytoma vs. Anaplastic Astrocytoma
res_LGA_APA <- as.data.frame(results(dds_history, contrast = c("history", "astrocytomas", "anaplastic astrocytomas")))
sig_LGA_APA <- res_LGA_APA[!is.na(res_LGA_APA$padj) & res_LGA_APA$padj < 0.05, ]

#Low Grade Astrocytoma vs. Primary Glioblastoma
res_LGA_GBM <- as.data.frame(results(dds_history, contrast = c("history", "astrocytomas", "primary Glioblastomas")))
sig_LGA_GBM <- res_LGA_GBM[!is.na(res_LGA_GBM$padj) & res_LGA_GBM$padj < 0.05, ]

#Low Anaplastic Astrocytoma vs. Primary Glioblastoma
res_APA_GBM <- as.data.frame(results(dds_history, contrast = c("history", "anaplastic astrocytomas", "primary Glioblastomas")))
sig_APA_GBM <- res_APA_GBM[!is.na(res_APA_GBM$padj) & res_APA_GBM$padj < 0.05, ]
```

This chunk adds a column called "SYMBOL", which has information about
the gene that is encoded by this cryptic ENSEBL ID.

```{r}
library(org.Hs.eg.db)
library(clusterProfiler)
library(AnnotationDbi)
sig_LGA_APA.df <- as.data.frame(sig_LGA_APA)
genes_clean <- sub("\\.\\d+$", "", rownames(sig_LGA_APA.df))
sig_LGA_APA.df$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
sig_LGA_APA.df
#PHLDA1-AS1 is 22 fold higher expressed in low-grade astrocytomas --> PHLDA1 protein is a target of the tumor suppressor gene p53 and helps to suppress tumors by inhibiting the Akt pathway


sig_LGA_GBM.df <- as.data.frame(sig_LGA_GBM)
genes_clean <- sub("\\.\\d+$", "", rownames(sig_LGA_GBM.df))
sig_LGA_GBM.df$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
sig_LGA_GBM.df
#BARX1 is 2.7 fold higher expressed in low-grade astrocytomas -->  Studies of the mouse and chick homolog suggest the encoded protein may play a role in developing teeth and craniofacial mesenchyme of neural crest origin.
#RCAN1 is 1.37 fold higher expressed in primary Gliobalstoma --> RCAN1 is an inhibitor of Calcineurin. Calcineurin is a key activator in T-Cell activation. Inhibition of it leads to immune supression
#IGF2 is 2 fold higher expressed in astrocytomas --> is an embryonal growth factor; might play a role in tumor progession



sig_APA_GBM.df <- as.data.frame(sig_APA_GBM)
genes_clean <- sub("\\.\\d+$", "", rownames(sig_APA_GBM.df))
sig_APA_GBM.df$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
sig_APA_GBM.df
#HGF is 2.3 times higher expressed in anaplastic-astrocytomas --> is a hematopoietic growth factor
#PHLDA1-AS1 is 22 fold higher expressed in primary glioblastomas --> PHLDA1 protein is a target of the tumor suppressor gene p53 and helps to suppress tumors by inhibiting the Akt pathway


```

## Not so important to be honest ##
## ________________________________ ##

This is a list of ALL genes that were found during comparison of the
three histories.

```{r}
DEG_all_astrocytomas <- unique(c(rownames(sig_LGA_APA), rownames(sig_LGA_GBM), rownames(sig_APA_GBM)))
DEG_all_astrocytomas
```

This is a table of the mean expression levels of the genes found in the
DEG-history analysis for the three typers of astrocytomas

```{r}
DEG_expr <- data.frame(matrix(0, 
                        nrow = length(DEG_all_astrocytomas), 
                        ncol = length(c("astrocytomas", "anaplastic astrocytomas", "primary Glioblastomas")),
                        dimnames = list(DEG_all_astrocytomas, c("astrocytomas", "anaplastic astrocytomas", "primary Glioblastomas"))),
  check.names = FALSE)

for (tumor in c("astrocytomas", "anaplastic astrocytomas", "primary Glioblastomas")){
  DEG_expr[, tumor] <- rowMeans(expr[DEG_all_astrocytomas, meta_data[meta_data$history %in% tumor, ]$Run])
}
DEG_expr 
```

```{r}
#Safes the list
#write.csv2(DEG_expr, "DEG_expr.csv", row.names = TRUE)
```

## ------------------------------------------------------------------------ ##

# Visualization of the expression levels of DEGs #

```{r}
library(ggplot2)
TOI <- c("astrocytomas", "anaplastic astrocytomas", "primary Glioblastomas")
expr_1 <- as.data.frame(meta_data[meta_data$history %in% TOI, c("Run", "history")])
```

## Boxplot of expression levels of the DEG-history genes ##

```{r}
library(ggplot2)

for (gene in DEG_all_astrocytomas) {
  b <- ggplot(expr_1, aes(x = history, y = .data[[gene]], fill = history)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2, alpha = 0.4) +
  theme_bw() +
  labs(title = gene, x = "Tumor Type", y = "Expression Level")

  print(b)
}

```

## Violin plot of expression levels of the DEG-history genes ##

```{r}
library(ggplot2)

for (gene in DEG_all_astrocytomas) {
  v <- ggplot(expr_1, aes(x = history, y = .data[[gene]], fill = history)) +
  geom_violin(trim = FALSE, alpha = 0.6) +
  geom_boxplot(width = 0.1, fill = "white") +
  theme_bw()


  print(v)
}

```

# Gene Ontology of the DEG genes #

```{r}
library(org.Hs.eg.db)
library(clusterProfiler)
library(AnnotationDbi)
```

## Creation of Background reference ##

## Different comparisons ##
###(Low-grade) Astrocytomas vs. anaplastic Astrocytomas ###


```{r}
#Create a DF from the significantly DEGs
sig_LGA_APA.df <- as.data.frame(sig_LGA_APA)

#Remove the ENSEBLE letters from the gene IDs
genes_clean <- sub("\\.\\d+$", "", rownames(sig_LGA_APA.df))

#Add the gene name for every gene
sig_LGA_APA.df$symbol <- mapIds(org.Hs.eg.db, keys = genes_clean, keytype = "ENSEMBL", column = "SYMBOL")
sig_LGA_APA.df

#Do the enrichment analysis
GO_LGA_APA <- enrichGO(gene = genes_clean, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
```

```{r}
#Dataframe of the enrighment GO analysis
as.data.frame(GO_LGA_APA[GO_LGA_APA$p.adjust < 0.005])
```

###(Low-grade) Astrocytomas vs. primary Glioblastomas ###

### Anaplastic Astrocytomas vs. primary Glioblastomas ###

To be fair, GO does not work with such a low amount of genes. We might
do that to compare astrocytomas to oligodendrocytomas or so. I think we
should rather have a manual look into the genes.
